src/ol_infrastructure/lib/fastly.py

```diff
@@ -6,6 +6,44 @@
 
 from bridge.secrets.sops import read_yaml_secrets
 
+__base_fastly_log_format_string = """{
+"client_ip":"%{json.escape(req.http.Fastly-Client-IP)}V",
+"fastly_is_edge":%{if(fastly.ff.visits_this_service==0,"true","false")}V,
+"fastly_server":"%{json.escape(server.identity)}V",
+"geo_city":"%{json.escape(client.geo.city)}V",
+"geo_conn_speed":"%{json.escape(client.geo.conn_speed)}V",
+"geo_conn_type":"%{json.escape(client.geo.conn_type)}V",
+"geo_continent_code":"%{json.escape(client.geo.continent_code)}V",
+"geo_country":"%{json.escape(client.geo.country_name)}V",
+"geo_country_code":"%{json.escape(client.geo.country_code)}V",
+"geo_region":"%{json.escape(client.geo.region.utf8)}V",
+"host":"%{if(req.http.Fastly-Orig-Host,json.escape(req.http.Fastly-Orig-Host),json.escape(req.http.Host))}V",
+"request_body_size_bytes":%{req.body_bytes_read}V,
+"request_duration_usec":%{time.elapsed.usec}V,
+"request_header_size_bytes":%{req.header_bytes_read}V,
+"request_method":"%{json.escape(req.method)}V",
+"request_protocol":"%{json.escape(req.proto)}V",
+"request_referer":"%{json.escape(req.http.referer)}V",
+"request_user_agent":"%{json.escape(req.http.User-Agent)}V",
+"response_body_size_bytes":%{resp.body_bytes_written}V,
+"response_header_size_bytes":%{resp.header_bytes_written}V,
+"response_reason":%{if(resp.response,"%22"+json.escape(resp.response)+"%22","null")}V,
+"response_state":"%{json.escape(fastly_info.state)}V",
+"response_status":%{resp.status}V,
+"timestamp":"%{strftime(\\{"%Y-%m-%dT%H:%M:%S%z"\\},time.start)}V",
+"url":"%{json.escape(req.url)}V"
+}"""
+
+
+# A fastly logformat string isn't actually json and we can't treat it as such in code
+# but we do need it to *produce* valid and minimized json at the end of the day once
+# it is installed into the fastly log configurations.
+def build_fastly_log_format_string(additional_static_fields: dict[str, str]) -> str:
+    split_format_string = __base_fastly_log_format_string.split("\n")
```

Comment:
Is this because it's not possible to treat the format string as valid JSON due to the special characters?

Reply:
Yeah I tried changing this to just a regular python-sh data structure and then using json.dumps but fastly refused to accept it as a valid logformat string. I didn't want to spend too much energy on it because I figured it didn't hurt to much to have it as just one big string above and the do some cleverness in this function to add additional fields while keeping the format string still producing valid json.